### PR TITLE
[WIP] Basemap

### PR DESCRIPTION
**Pull request by nerik**
Update to the Basemap Generator to handle several layers - layer composition is handled by the `sort` transformation.
Screenshot 2020-02-08 at 17 14 52

Additionally, generators now have separated GeneratorConfig types extending a generic type.

TODO: 
- [x] Update generated tiles to a bucket
- [ ] Update CP
- [ ] Attribution functionality has been removed: it will be implemented later by the `addAttributions` transformation (collect all attribution across all layers, uniq them, concat them, add them to global style metadata as a string/html)

### CODE REVIEW

File: scripts/server.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env node
+
+const path = require('path')
+const express = require('express')
+const app = express()
+const cors = require('cors')
+
+app.use(cors())
+app.use(
```
Comment (j8seangel) [inline scripts/server.js:11]: is this just for testing locally purpose ?

File: src/layer-composer/generators/background/background.ts [hunk 2]
```diff
@@ -4,6 +4,10 @@ import { GeneratorConfig } from 'layer-composer/types'
 
 export const BACKGROUND_TYPE = 'BACKGROUND'
 
+export interface BackgroundGeneratorConfig extends GeneratorConfig {
```
Comment (j8seangel) [inline src/layer-composer/generators/background/background.ts:7]: I like this idea, so why not having the `type` defined here too?

File: src/layer-composer/generators/basemap/basemap-layers.ts [hunk 3]
```diff
@@ -1,27 +1,73 @@
...
+    source: BASEMAPS.Satellite,
+    metadata: {
+      group: Group.Basemap,
+    },
+  },
+  [BASEMAPS.Landmass]: {
+    type: 'fill',
+    id: BASEMAPS.Landmass,
+    source: BASEMAP_VECTOR_SOURCE,
+    'source-layer': 'landmass',
```
Comment (j8seangel) [inline src/layer-composer/generators/basemap/basemap-layers.ts]: Why are you duplicating the string here?
```suggestion
    'source-layer':  BASEMAPS.Landmass,
```

File: src/layer-composer/generators/basemap/basemap-layers.ts [hunk 4]
```diff
@@ -1,27 +1,73 @@
...
+      'https://mt1.google.com/vt/lyrs=s&x={x}&y={y}&z={z}',
+      'https://mt2.google.com/vt/lyrs=s&x={x}&y={y}&z={z}',
+      'https://mt3.google.com/vt/lyrs=s&x={x}&y={y}&z={z}',
+    ],
+    type: 'raster',
+    tileSize: 256,
+  },
+  [BASEMAP_VECTOR_SOURCE]: {
+    type: 'vector',
+    tiles: ['http://localhost:9090/tiles-basemap/{z}/{x}/{y}.pbf'],
```
Comment (j8seangel) [inline src/layer-composer/generators/basemap/basemap-layers.ts]: Would be nice to have a `tiles` params in the basemapGeneratorConfig to be able to resuse this

File: src/layer-composer/generators/background/background.ts (hunk 2 above)
Comment (nerik) [reply src/layer-composer/generators/background/background.ts:7]: I don't think that's possible :/

File: scripts/server.js (hunk 1 above)
Comment (nerik) [reply scripts/server.js:11]: Yes, added a comment

File: src/layer-composer/generators/basemap/basemap-layers.ts (hunk 4 above)
Comment (nerik) [reply src/layer-composer/generators/basemap/basemap-layers.ts]: I'm not sure. If we want this we might prefer having a more generic `VectorTilesGenerator` or something. I think the Basemap is a thing we want to drop in any project and forget about it, not something that should be configurable.
Let's reasses when we have a use case for this.

**Comment by nerik** [comment]
@j8seangel thx for the review!

**Pull request merged**